src/runtime_strings_extra.js

```diff
@@ -117,16 +120,18 @@ function lengthBytesUTF16(str) {
   return str.length*2;
 }
 
-function UTF32ToString(ptr) {
+function UTF32ToString(ptr, maxBytesToRead) {
 #if ASSERTIONS
   assert(ptr % 4 == 0, 'Pointer passed to UTF32ToString must be aligned to four bytes!');
 #endif
   var i = 0;
 
   var str = '';
-  while (1) {
+  // If maxBytesToRead is not passed explicitly, it will be undefined, and this
+  // will always evaluate to true. This saves on code size.
+  while (!(i >= maxBytesToRead / 4)) {
     var utf32 = {{{ makeGetValue('ptr', 'i*4', 'i32') }}};
```

Comment:
These makeGetValues could be demoted to direct HEAPU32/HEAPU16 reads for better performance and smaller code size, although could leave that for a later day as well.

Reply:
The compiler turns those `{{{ makeGetValue }}}` things into `HEAP32` at compile time, so there should be no speed or size difference.

Reply:
I don't think it is able to optimize away the *4, so it will end up with a *4 in each iteration. Manually iterating over indices instead of iterating over bytes will be shorter code size wise and also faster too.

Reply:
That's a good question, yeah, maybe not. Sounds good to iterate over `HEAP32` indexes in a followup.